src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs

```diff
@@ -492,5 +497,120 @@ private static unsafe void ArrayTypeCheck_Helper(object obj, void* elementType)
                 ThrowArrayMismatchException();
             }
         }
+
+        // Helpers for Unboxing
+        [DebuggerHidden]
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static unsafe void InitValueClass(ref byte destBytes, MethodTable *pMT)
+        {
+            uint numInstanceFieldBytes = pMT->GetNumInstanceFieldBytes();
+            if ((((uint)Unsafe.AsPointer(ref destBytes) | numInstanceFieldBytes) & ((uint)sizeof(void*) - 1)) != 0)
```

Comment:
Is is more or less efficient to check `pMT->ContainsGCPointers`?

Reply:
I'm not sure. I'm not quite ready to benchmark this, and need to validate that we actually need to do aligned work here at all. I see 3 basic implementations.
1. This bifurcated approach, with some sort of if check on safety of using the general purpose clearing routine. That sort of thing is safe for cases where the destination may be on the heap or not. Using `ContainsGCPointers` wouldn't have been safe for the C++ implementation of all of this, but since this implementation of InitValueClass is actually only used for `Nullable<T>` instances, we CAN use that flag. I need to some performance investigation to determine the actual fastest approach here.
2. An approach where we p/invoke to 'C' memset as an FCall. That is only safe if the destination is guaranteed to be on the current stack. This needs to be an FCall since otherwise the specification of memset permits non-pointer atomic stores, and an inopportune GC suspension could see an invalid pointer.
3. A custom set routine which does a warmup series of sets to individual bytes until the region is aligned, then sets pointer sized chunks using atomic stores. (The CPU notion of atomic, not the C++ notion), and then has a warmdown phase. This is safe for all cases as well, and is what the old C++ implementation used to do.

A similar set of choices exists for the actual copying routine below.

Reply:
In any case, my analysis of the usage of Unbox_Nullable is that usage for the JIT scenario always works with a stack based destination, but there is usage from Array.CoreCLR.cs which can unbox to the heap. I'm building a small suite of perf tests to see what the performance of various options in this space looks like.